### PR TITLE
Create clickable links from the listed context files

### LINKED ISSUE #255: Make contextFiles a clickable link

### PR DESCRIPTION
**Pull request by mkondratek**
Fixes https://github.com/sourcegraph/jetbrains/issues/255

## Test plan 
- `:runIde`
- ask Cody about the code / run explain command 
- see the context file
- text the links

image

Note that we have issue with unrelated files shown in the context read files list: https://github.com/sourcegraph/jetbrains/issues/298

### CODE REVIEW

File: src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    this.layout = BorderLayout()
+
+    val margin = JBInsets.create(Insets(TEXT_MARGIN, TEXT_MARGIN, TEXT_MARGIN, TEXT_MARGIN))
+    val contextFileNames =
+        contextMessages
+            .stream()
+            .map<ContextFile>(ContextMessage::file)
+            .filter { obj: ContextFile? -> Objects.nonNull(obj) }
+            .map(ContextFile::fileName)
+            .collect(Collectors.toSet())
```
Comment (exigow) [inline src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt]: We can remove Java stream & collect.

File: src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    contextFileNames.forEach { fileName: String ->
+      val filePanel = createFileWithLinkPanel(project, fileName)
+      accordionSection.contentPanel.add(filePanel, fileIndex.getAndIncrement())
+    }
+    add(accordionSection, BorderLayout.CENTER)
+  }
+
+  private fun createFileWithLinkPanel(project: Project, fileName: String): JPanel {
+    val fileWithoutRedundantPrefix =
+        fileName.removePrefix("../../../..").removePrefix(project.baseDir.path)
```
Comment (exigow) [inline src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt]: Can you pls elaborate on this `removePrefix("../../../..")`?  It looks like a underlying problem. Are we getting such paths from the agent?

File: src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt (hunk 1 above)
Comment (exigow) [reply src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt]: plus use filterNotNull + toSet()

File: src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    val anAction =
+        object : DumbAwareAction() {
+          override fun actionPerformed(anActionEvent: AnActionEvent) {
+            val file = project.baseDir.findFileByRelativePath(fileWithoutRedundantPrefix)
```
Comment (exigow) [inline src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt]: Looks like `com.intellij.openapi.project.Project.getBaseDir()` is deprecated.

Maybe `getBasePath()` is better alternative?

File: src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            logger.info("Opening a file from the used context (fileName=$fileName file=$file)")
+            if (file != null) {
+              FileEditorManager.getInstance(project).openFile(file, /*focusEditor=*/ true)
```
Comment (exigow) [inline src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt]: IMO we can ~remove this `/*focusEditor=*/` or~ replace with `val focusEditor = true`.

File: src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+            .stream()
+            .map<ContextFile>(ContextMessage::file)
+            .filter { obj: ContextFile? -> Objects.nonNull(obj) }
+            .map(ContextFile::fileName)
+            .collect(Collectors.toSet())
+
+    val accordionSection = AccordionSection("Read ${contextFileNames.size} files")
+    accordionSection.isOpaque = false
+    accordionSection.border = EmptyBorder(margin)
+    val fileIndex = AtomicInteger(0)
```
Comment (exigow) [inline src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt]: Suggestion: `AtomicInteger` -> `Int`

(code is synchronous)

Comment (exigow) [reply src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt]: Or even better: `forEachIndexed`! :smile:

**Review by exigow — CHANGES_REQUESTED** [review]
Looks good! :+1: I've requested some small changes (see: comments).

File: src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt (hunk 1 above)
Comment (mkondratek) [reply src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt]: what can I do instead?

Comment (mkondratek) [reply src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt]: got it, act

File: src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt (hunk 4 above)
Comment (mkondratek) [reply src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt]: We cannot do any better here.

File: src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt (hunk 5 above)
Comment (mkondratek) [reply src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt]: ok

File: src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt (hunk 2 above)
Comment (mkondratek) [reply src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt]: nvm, I was using a wrong repo as a context

File: src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt (hunk 3 above)
Comment (mkondratek) [reply src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt]: thanks!

Comment (pkukielka) [inline src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt]: Wasn't `findFileByRelativePath` one of the methods which sometimes were taking longer than expected?
If that is the case, remember we should use the same constraints for the listener callbacks and actions as we use for UI threads - this means no long running actions. Just FYI.

File: src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt (hunk 2 above)
Comment (pkukielka) [inline src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt]: That looks weird. Why exactly `../../../..`?

**Review by exigow — APPROVED** [review]
Thanks for the fixes.

File: src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt (hunk 2 above)
Comment (mkondratek) [reply src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt]: or... there is a problem 🤔

Comment (mkondratek) [reply src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt]: image

installed on my IDE & run in jetbrains project

File: src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt (hunk 3 above)
Comment (mkondratek) [reply src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt]: nah, it was `getRelativePathToProjectRoot`. but I will take a close look 🕵️

Comment (mkondratek) [reply src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt]: fixed

File: src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt (hunk 2 above)
Comment (mkondratek) [reply src/main/kotlin/com/sourcegraph/cody/chat/ContextFilesMessage.kt]: nah, I probably had the former bad version or bad context repo - it works properly now.

there is another problem, however - I will start a discussion below

**Comment by mkondratek** [comment]
So the remaining unknown seems to be the remote files. Take a look at the agent's response:

image

It contains a doubled entry of the local file:
```
{
  "fileName": "/Users/mkondratek/IdeaProjects/runIde/jetbrains/src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt"
},
```
and two other entries with repo & revision:
```
{
  "fileName": "src/main/kotlin/com/sourcegraph/cody/chat/CodyOnboardingGuidancePanel.kt",
  "repoName": "github.com/sourcegraph/jetbrains",
  "revision": "14891e64eb192789df371dd43ae819a77aa22cd6"
},
{
  "fileName": "src/main/java/com/sourcegraph/cody/CodyToolWindowContent.java",
  "repoName": "github.com/sourcegraph/jetbrains",
  "revision": "14891e64eb192789df371dd43ae819a77aa22cd6"
},
```

I assume that the file referencing the local file should link it directly. 

What about the "remote files"? We can link the local files to them BUT there can be inconsistencies b/w local and remote (eg. the agent uses a file that exists in the remote but it does not exist locally). How should we handle it?

File: gradle.properties [hunk 6]
```diff
@@ -10,12 +10,12 @@ pluginSinceBuild=221.1
 pluginUntilBuild=*
 # IntelliJ Platform Properties -> https://github.com/JetBrains/gradle-intellij-plugin#intellij-platform-properties
 platformType=IC
-platformVersion=2022.1
+platformVersion=2023.3.1
```
Comment (pkukielka) [inline gradle.properties]: Rollback this please!

**Review by pkukielka — CHANGES_REQUESTED** [review]
Please do not bump IJ platform version!

**Comment by exigow** [comment]
@danielmarquespt, we haven't consulted this change with you so please take a look on attached demo in first comment & eventually confirm. The design has slightly changed (paths now have "link" color + no border/outline).

Before:
image

**Comment by mkondratek** [comment]
it should be all good now

**Comment by mkondratek** [comment]
> @danielmarquespt, we haven't consulted this change with you so please take a look on attached demo in first comment & eventually confirm. The design has slightly changed (paths now have "link" color + no border/outline).
> 
> Before: image

image

**Review by pkukielka — APPROVED** [review]
LGTM

**Pull request merged**